### PR TITLE
Add check to ensure locked dependencies have source distributions available.

### PR DESCRIPTION
**Pull request by reivilibre**
It'd be nice if it was obvious when the Poetry lockfile has not been updated in a safe way.
Crucially, all dependencies need an sdist or else we won't be able to install them everywhere.

Follows: #14718 

Base: `develop`

Original commit schedule, with full messages:

<ol>
<li>

Add a script to check the lockfile for dependencies without sdists. 

</li>
<li>

Add dependency on tomli 

</li>
<li>

Add workflow to ensure locked deps have sdists 

</li>
</ol>

### CODE REVIEW

File: poetry.lock [hunk 1]
```diff
@@ -1967,6 +1970,7 @@ lxml = [
     {file = "lxml-4.9.2-cp310-cp310-musllinux_1_1_x86_64.whl", hash = "sha256:b9cc34af337a97d470040f99ba4282f6e6bac88407d021688a5d585e44a23184"},
     {file = "lxml-4.9.2-cp310-cp310-win32.whl", hash = "sha256:d02a5399126a53492415d4906ab0ad0375a5456cc05c3fc0fc4ca11771745cda"},
     {file = "lxml-4.9.2-cp310-cp310-win_amd64.whl", hash = "sha256:a38486985ca49cfa574a507e7a2215c0c780fd1778bb6290c21193b7211702ab"},
+    {file = "lxml-4.9.2-cp311-cp311-macosx_10_15_universal2.whl", hash = "sha256:6943826a0374fb135bb11843594eda9ae150fba9d1d027d2464c713da7c09afe"},
```
Comment (DMRobertson) [inline poetry.lock:1973]: Surprised that this wheel now shows up!

File: poetry.lock [hunk 2]
```diff
@@ -1977,7 +1981,6 @@ lxml = [
     {file = "lxml-4.9.2-cp35-cp35m-manylinux_2_5_i686.manylinux1_i686.whl", hash = "sha256:ca989b91cf3a3ba28930a9fc1e9aeafc2a395448641df1f387a2d394638943b0"},
     {file = "lxml-4.9.2-cp35-cp35m-manylinux_2_5_x86_64.manylinux1_x86_64.whl", hash = "sha256:822068f85e12a6e292803e112ab876bc03ed1f03dddb80154c395f891ca6b31e"},
     {file = "lxml-4.9.2-cp35-cp35m-win32.whl", hash = "sha256:be7292c55101e22f2a3d4d8913944cbea71eea90792bf914add27454a13905df"},
-    {file = "lxml-4.9.2-cp35-cp35m-win_amd64.whl", hash = "sha256:998c7c41910666d2976928c38ea96a70d1aa43be6fe502f21a651e17483a43c5"},
```
Comment (DMRobertson) [inline poetry.lock:1980]: Surprised that this wheel has gone

File: poetry.lock [hunk 3]
```diff
@@ -1987,7 +1990,6 @@ lxml = [
     {file = "lxml-4.9.2-cp36-cp36m-musllinux_1_1_aarch64.whl", hash = "sha256:58bfa3aa19ca4c0f28c5dde0ff56c520fbac6f0daf4fac66ed4c8d2fb7f22e74"},
     {file = "lxml-4.9.2-cp36-cp36m-musllinux_1_1_x86_64.whl", hash = "sha256:bc718cd47b765e790eecb74d044cc8d37d58562f6c314ee9484df26276d36a38"},
     {file = "lxml-4.9.2-cp36-cp36m-win32.whl", hash = "sha256:d5bf6545cd27aaa8a13033ce56354ed9e25ab0e4ac3b5392b763d8d04b08e0c5"},
-    {file = "lxml-4.9.2-cp36-cp36m-win_amd64.whl", hash = "sha256:3ab9fa9d6dc2a7f29d7affdf3edebf6ece6fb28a6d80b14c3b2fb9d39b9322c3"},
```
Comment (DMRobertson) [inline poetry.lock:1990]: Also surprised that this wheel has gone

File: poetry.lock [hunk 4]
```diff
@@ -2292,12 +2294,34 @@ psycopg2cffi-compat = [
...
+    {file = "pyasn1_modules-0.2.8-py2.7.egg", hash = "sha256:0845a5582f6a02bb3e1bde9ecfc4bfcae6ec3210dd270522fee602365430c3f8"},
     {file = "pyasn1_modules-0.2.8-py2.py3-none-any.whl", hash = "sha256:a50b808ffeb97cb3601dd25981f6b016cbb3d31fbf57a8b8a87428e6158d0c74"},
+    {file = "pyasn1_modules-0.2.8-py3.1.egg", hash = "sha256:f39edd8c4ecaa4556e989147ebf219227e2cd2e8a43c7e7fcb1f1c18c5fd6a3d"},
+    {file = "pyasn1_modules-0.2.8-py3.2.egg", hash = "sha256:b80486a6c77252ea3a3e9b1e360bc9cf28eaac41263d173c032581ad2f20fe45"},
+    {file = "pyasn1_modules-0.2.8-py3.3.egg", hash = "sha256:65cebbaffc913f4fe9e4808735c95ea22d7a7775646ab690518c056784bc21b4"},
+    {file = "pyasn1_modules-0.2.8-py3.4.egg", hash = "sha256:15b7c67fabc7fc240d87fb9aabf999cf82311a6d6fb2c70d00d3d0604878c811"},
+    {file = "pyasn1_modules-0.2.8-py3.5.egg", hash = "sha256:426edb7a5e8879f1ec54a1864f16b882c2837bfd06eee62f2c982315ee2473ed"},
+    {file = "pyasn1_modules-0.2.8-py3.6.egg", hash = "sha256:cbac4bc38d117f2a49aeedec4407d23e8866ea4ac27ff2cf7fb3e5b570df19e0"},
+    {file = "pyasn1_modules-0.2.8-py3.7.egg", hash = "sha256:c29a5e5cc7a3f05926aff34e097e84f8589cd790ce0ed41b67aed6857b26aafd"},
```
Comment (DMRobertson) [inline poetry.lock:2324]: I think these eggs are unrelated (no idea why they only show up now).

Can you try nuking your poetry caches and re-invoke `poetry lock --no-update` and see if the lockfile gets cleaner?

File: pyproject.toml [hunk 5]
```diff
@@ -352,6 +352,9 @@ twine = "*"
 # Towncrier min version comes from #3425. Rationale unclear.
 towncrier = ">=18.6.0rc1"
 
+# Used for checking the Poetry lockfile
+tomli = ">=1.2.3"
+
```
Comment (DMRobertson) [inline pyproject.toml:357]: Note that `tomli` wasn't added to the lockfile; presumably it was already pulled in by something else?

File: .github/workflows/poetry_lockfile.yaml [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+  pull_request:
+    paths:
+      - poetry.lock
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
+permissions:
+  contents: write
```
Comment (DMRobertson) [inline .github/workflows/poetry_lockfile.yaml:15]: Is this correct? I would have thought the script only reads

File: scripts-dev/check_locked_deps_have_sdists.py [hunk 7]
```diff
@@ -24,9 +24,16 @@ def main() -> None:
...
+    packages_to_assets: Dict[str, List[Dict[str, str]]]
+    # There seem to be two different formats for storing the list of files per package.
+    if lockfile_content.get("metadata", {}).get("files") is not None:
+        # either [metadata.files]
+        packages_to_assets = lockfile_content["metadata"]["files"]
+    else:
+        # or a `files` inline table in each [[package]]
+        packages_to_assets = {
+            package["name"]: package["files"] for package in lockfile_content["package"]
+        }
```
Comment (DMRobertson) [inline scripts-dev/check_locked_deps_have_sdists.py:36]: Seems to be part of the poetry 1.3 lockfile format, see https://github.com/python-poetry/poetry/pull/6393

**Review by DMRobertson — CHANGES_REQUESTED** [review]
PTAL at

- Permissions on the action
- Unexpected guff in the lockfile

the other comments are just me thinking out loud.

File: scripts-dev/check_locked_deps_have_sdists.py (hunk 7 above)
Comment (DMRobertson) [reply scripts-dev/check_locked_deps_have_sdists.py:36]: Note that I'm about to require the new lockfile format in https://github.com/matrix-org/sytest/pull/1327

**Comment by clokep** [comment]
> Crucially, all dependencies need an sdist or else we won't be able to install them everywhere.

Is it correct that "everywhere" means:

* If you use a Python distribution that doesn't have wheels, e.g. Pyston or brand new Pythons (Python 3.11)
* If you run on hardware that doesn't have wheels for some things (ARM? Raspberry Pi?)

I'm sure there are other examples too.

**Comment by reivilibre** [comment]
> > Crucially, all dependencies need an sdist or else we won't be able to install them everywhere.
> 
> Is it correct that "everywhere" means:
> 
>     * If you use a Python distribution that doesn't have wheels, e.g. Pyston or brand new Pythons (Python 3.11)
> 
>     * If you run on hardware that doesn't have wheels for some things (ARM? Raspberry Pi?)
> 
> 
> I'm sure there are other examples too.

Yeah, both of those would apply. Note that 'Brand new Pythons' can still be OK with wheels that predate them if the wheels are targetted at ABI3, but different CPU architectures is a notable one. (Also, some distributions might refuse to use, or be unable to use, wheels.)

File: .github/workflows/poetry_lockfile.yaml [hunk 8]
```diff
@@ -0,0 +1,24 @@
+on:
+  push:
+    branches: ["develop", "release-*"]
+    paths:
+      - poetry.lock
+  pull_request:
+    paths:
+      - poetry.lock
```
Comment (DMRobertson) [inline .github/workflows/poetry_lockfile.yaml:8]: Nice that we only trigger this if needed.

The downside is that it can't be fed into the `tests-done` job to tell github it's safe to merge this PR. But I think we can live with that.

**Review by DMRobertson — APPROVED** [review]

File: .github/workflows/poetry_lockfile.yaml (hunk 8 above)
Comment (reivilibre) [reply .github/workflows/poetry_lockfile.yaml:8]: It's frustrating how the default behaviour isn't just to make sure all the tests pass, but ah well
